plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -58,37 +56,35 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
         }
 
     override fun visitWhenExpression(whenExpression: FirWhenExpression, data: StmtConversionContext): Exp {
-        if (whenExpression.usedAsExpression) {
-            // The problem with conditional expressions is that Kotlin conditional expressions can have statements whereas Viper
-            // conditionals can't. These need to be embedded as Viper statements before the IF, however, one has to carefully consider
-            // short-circuiting semantics.
-            TODO("Conditionals used as expressions are not yet implemented.")
-        }
-
         if (whenExpression.branches.size != 2) {
             // For now only when expressions with 2 branches are supported so if statements can be encoded.
             // When expressions with more than 2 branches can be embedded as nested if-else chains.
             TODO("When expressions with more that two branches are not yet implemented")
         }
 
-        when (whenExpression.subject) {
+        return when (whenExpression.subject) {
             null -> {
+                val cvar = data.methodCtx.newAnonVar(data.methodCtx.programCtx.convertType(whenExpression.typeRef.coneTypeOrNull!!))
                 val cond = whenExpression.branches[0].condition.accept(this, data)
+                // TODO: tidy this up and split it into helper functions.
                 val thenCtx = StmtConversionContext(data.methodCtx)
-                thenCtx.convertAndAppend(whenExpression.branches[0].result)
+                val thenResult = whenExpression.branches[0].result.accept(this, thenCtx)
+                thenCtx.statements.add(Stmt.LocalVarAssign(cvar.toLocalVar(), thenResult))
                 val elseCtx = StmtConversionContext(data.methodCtx)
-                elseCtx.convertAndAppend(whenExpression.branches[1].result)
+                val elseResult = whenExpression.branches[1].result.accept(this, elseCtx)
+                elseCtx.statements.add(Stmt.LocalVarAssign(cvar.toLocalVar(), elseResult))
+                data.declarations.add(cvar.toLocalVarDecl())
                 data.statements.add(
                     Stmt.If(
                         cond,
                         thenCtx.block,
                         elseCtx.block
                     )
                 )
+                cvar.toLocalVar()
```

Comment:
Maybe we can wrap line 67, 72, 75 and 84 into an `if (whenExpression.usedAsExpression)` block to not unnecessarily bloat the Viper output if we can easily fix it.

Reply:
I'm reluctant to do that right now since the control flow ends up quite convoluted, and it's not clear this code will stay relevant for long. Our approach with anonymous variables is going to be expensive regardless; if it turns out to be a problem, a more structured solution is probably called for.

Reply:
Alright, let's keep it like this for now.